platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java

```diff
@@ -250,19 +258,32 @@ void add(T element) {
             throw new IllegalStateException("You can not call add on a immutable ConcurrentArray");
         }
 
-        // The sub-arrays in the dequeue all considered full except for the very last one. If the very last one is
+        // Lazily initialize the head of this copy.
+        if (head == null) {
```

Comment:
This is Double-Checked-Locking, which results in incorrect behavior in java.
You will always need the memory barrier to make this correct when called from multiple threads.
Options are to just skip the first test (i.e. `synchronized {` on this line without the outer `if`) or, if seeking to improve performance, use a ReentrantLock instead of the Object monitor lock implied by `synchronized`.

Reply:
`head` is volatile...

Reply:
I reimplemented the linked list in `ConcurrentArray` due to a design flaw found by Artem.
Java fixed Double-Checked-Locking with `volatile`, though.